docs/proposals/project-repos-and-clusters.md

```diff
@@ -0,0 +1,170 @@
+We want to streamline the process of adding repositories and clusters to the project and make it self-service.
+
+## Motivation
+
+As long as the developer has the required credentials he/she should be able to add repository/cluster to the project
+without involving the administrator. To archive it, we are proposing to introduce project scoped repositories and clusters.
+
+### Goals
+
+The goals of project scoped repositories and clusters are:
+
+#### Allow Self-Registering Repositories/Clusters in a Project
+
+Developer should be able to add a repository/cluster into the project without asking help from Argo CD administrator.
+
+### Non-Goals
+
+#### Simplify Management of Shared Repositories/Clusters in a Project
+
+The repositories and clusters that can be used across multiple projects still have to be managed by Argo CD administrator.
+
+## Proposal
+
+#### Project scoped repository/cluster
+
+The proposal is to introduce project scoped clusters and repositories that can be managed by a developer who has access to the project.
+The only difference of project scoped repository/cluster is that it has `project` field with the project name it belongs to. Both repositories
+and clusters are stored as Kubernetes Secrets, so a new field could be stored as a Secret data key:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: argocd-example-apps
+  labels:
+    argocd.argoproj.io/secret-type: repository
+type: Opaque
+stringData:
+  project: my-project1                                     # new project field
+  name: argocd-example-apps
+  url: https://github.com/argoproj/argocd-example-apps.git
+  username: ****
+  passoword: ****
+```
+
+* The project scoped repository/cluster is automatically allowed in the project.
+  This enables developers to allow new cluster/repository without modifying the project.
+* The project scoped repository/cluster still can be used in other project but it has to be allowed by admin (as normal repository/cluster).
+* If another team wants to add the same repository/cluster into a different project they would have to ask admin. 
+
+#### Project RBAC Changes
+
+The organization still might want to enforce certain rules so developers won't get permission to add a
+project-specific repository/cluster by default. The administrator might use RBAC to control access to
+the project scoped repositories cluster. The access to project scope actions will be checked using
+`<projectName>/<name>` pattern. For example, to allow users to add project scoped repositories admin would have to add
+the following RBAC rules:
+
+```
+p, proj:my-project:admin, repositories, create, my-project/*, allow
+p, proj:my-project:admin, repositories, delete, my-project/*, allow
+p, proj:my-project:admin, repositories, update, my-project/*, allow
+```
+
+This provides extra flexibility so that admin can have stricter rules. e.g.:
+
+```
+p, proj:my-project:admin, repositories, update, my-project/"https://github.my-company.com/*", allow
+```
```

Comment:
Great. I was wondering how we would enforce rbac.

Reply:
fixed